### PR TITLE
Actuator zone modifications and fixes

### PR DESCRIPTION
**Pull request by anilyil**
## Purpose

This PR adds the heat addition feature in the actuator region. Just like adding thrust equally among the actuator zone, we can now also specify an amount of heat power to be added to the flow. This BC variable is differentiated. There is an accompanying PR in baseclasses for the definition of the BC variable. See: https://github.com/mdolab/baseclasses/pull/44 

This also fixes a few more things:
- Bug fix in the differentiation of the flow power integration
- Added a few entries to the gitignore
- Fixed a few docstrings in pyadflow
- Fixed the indentation of the BC printout from Fortran level

I have also created several tests for the actuator zone. These are based on the initial efforts of @joanibal, so thank you Josh! 

The tests include analysis, adjoint, and partial checks. I also added complex tests, and verified the adjoint sensitivities with the complex step. The complex tests will not run right now because they need fixes for the ANK solver. I am creating this PR right now so people can review the changes, and once the complex ANK is also fixed (will do it soon), then we can merge this PR.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

## Testing

I added new real and complex tests. All tests pass for me locally.

## Checklist

- [ ] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Comment by anilyil** [comment]
I have converted this to a draft because the complex tests won't pass until I fix the complex ANK. Other than that, please take a look and give me feedback on the changes, so we can merge this after the complex solvers are fixed.

**Comment by anilyil** [comment]
I just remembered that I also need to add the files required for the tests to the inputfiles tarball. Can you help me with that step @nwu63 @eirikurj ?

**Comment by anilyil** [comment]
I have updated the input files, and the baseclasses PR is merged. As expected, the real tests pass now. The complex ones will also pass once we update the complex solvers.

**Comment by nwu63** [comment]
A reminder for @anilyil to bump the minor version here and make a release after this is merged.

**Comment by anilyil** [comment]
This one is ready for review @joanibal @sseraj

File: tests/reg_tests/test_actuator.py
Comment (joanibal) [inline tests/reg_tests/test_actuator.py:113]: If it tests both momentum and power we should change the test name

File: tests/reg_tests/test_actuator.py
Comment (joanibal) [inline tests/reg_tests/test_actuator.py:225]: This seems like a hack to me. 
Why not just use an absolute tolerance?

**Review by joanibal — CHANGES_REQUESTED** [review]
A few nitpicks

File: tests/reg_tests/test_actuator.py
Comment (sseraj) [inline tests/reg_tests/test_actuator.py:288]: "integrated"

File: tests/reg_tests/test_actuator.py
Comment (anilyil) [reply tests/reg_tests/test_actuator.py:113]: So I think I meant by test momentum flux, is to test adding thrust terms, and checking if it has the correct effect on both momentum and energy fluxes through the boundaries (you can add thrust in the actuator zone and it both adds momentum to the flow, and does work on the flow) 

I renamed the tests. Let me know if they are more clear.

File: tests/reg_tests/test_actuator.py
Comment (anilyil) [reply tests/reg_tests/test_actuator.py:225]: I don't know how to fix this, and I am probably being dumb but I also could not find a direct answer online. The issue is because the target is exactly zero. Then the relative error ends up being infinite:
```
AssertionError: 
Not equal to tolerance rtol=0.001, atol=0.001

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 0.07964251
Max relative difference: inf
 x: array(0.079643)
 y: array(0.)
```

Do you have any suggestions on how to fix this? I tried setting a rtol of `np.inf` but still fails. I am not sure if there is a version that just checks atol. Swapping the order of value and target also does not help, and that's actually not a great solution in general imo.

What I can do is set the target to the small 0.07964251 value. This is due to the accumulation of integration errors and flow convergence in ADflow. If I compare against this (nonzero) target, then both atol and rtol will work. Or I could add a finite number to both sides (100, or 1 or whatever). Let me know what you think is the best solution here. I am leaving this test in a broken state so we dont merge until this is fixed.

File: adflow/pyADflow.py
Comment (anilyil) [reply adflow/pyADflow.py:739]: Good catch. I fixed these in `writeActuatorRegions` and `addActuatorRegion`. I also opened an issue linking your comment. Let me know if there are other docstrings I should fix.

File: adflow/pyADflow.py
Comment (anilyil) [reply adflow/pyADflow.py:621]: Yeah good point, I missed this text here. I updated this part of the docstring completely to be more clear.

File: adflow/pyADflow.py
Comment (anilyil) [reply adflow/pyADflow.py:623]: Seems like this paragraph was mostly a mess. I updated this part of the docstring completely to be more clear.

File: adflow/pyADflow.py
Comment (anilyil) [reply adflow/pyADflow.py:622]: Updated that whole paragraph.

File: src/modules/actuatorRegionData.F90
Comment (anilyil) [reply src/modules/actuatorRegionData.F90:22]: yeah I agree. it seems we only use these inside the derived `actuatorRegionType`. I was not going to change this but then I realized we also have another %q in flowdoms for viscous surfaces, I am guessing for heat fluxes. I will change these to thrust, torque and heat. We still use them inside the derived data type so having the same parameter name should not affect it in this routine. Worst case, I will rename the inputs with `_in` or sth. I will do this change in a separate commit.

File: src/bcdata/BCData.F90
Comment (anilyil) [reply src/bcdata/BCData.F90:1585]: Yes, setPointer changes are. I fixed this while I was fixing the others. As far as I know, the current torque implementation  does not actually do anything. @shamsheersc19 fixed it at some point for his stuff but that was a separate work. We may look into merging his torque stuff here at some point.

File: src/solver/residuals.F90
Comment (anilyil) [reply src/solver/residuals.F90:352]: lol, I forgot to remove this. Probably used it while debugging. removing it.

File: src/solver/residuals.F90
Comment (anilyil) [reply src/solver/residuals.F90:318]: Because it seems this is defined as `LRef` in https://github.com/mdolab/adflow/blob/master/src/modules/flowVarRefState.F90#L43-L50 

I can change it to `lRef` if you want?

File: tests/reg_tests/test_actuator.py
Comment (anilyil) [reply tests/reg_tests/test_actuator.py:288]: fixed 3 instances of this typo. words are difficult

File: src/modules/actuatorRegionData.F90
Comment (anilyil) [reply src/modules/actuatorRegionData.F90]: See https://github.com/mdolab/adflow/pull/118/commits/b10c7226354e2350787011d5e484406e81eced8f for this fix. Let me know if the new names look ok.

Comment (anilyil) [reply src/modules/actuatorRegionData.F90]: See b10c722 for this fix. Let me know if the new names look ok.

**Comment by anilyil** [comment]
thanks for the review @joanibal and @sseraj. I think I addressed all of your comments except Josh's comment at: https://github.com/mdolab/adflow/pull/118#discussion_r599993034 Once we figure out that test, this is good to go for me.

File: tests/reg_tests/test_actuator.py
Comment (sseraj) [reply tests/reg_tests/test_actuator.py:225]: I think the `inf` is expected, it's just that the `atol` is too tight.

Comment (nwu63) [reply tests/reg_tests/test_actuator.py:225]: The function `numpy.testing.assert_allclose` that you used should do exactly what you want, is that not the case? The [docstrings](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) say it should compare using `atol + rtol * abs(desired)` so if `desired=0` then it should use just `atol`. Maybe you need a looser `atol` since the actual tolerance is around 0.08.

However, if you expect the integration error to be around 0.08 then that may be a better value to use. That will ensure a tighter tolerance and check if in the future this value changes by a small amount.

File: adflow/pyADflow.py
Comment (nwu63) [reply adflow/pyADflow.py]: Is this true? I haven't found any issues so far. Also, the [style guide](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) explicitly says to have a space before.

We could consider adopting a docstring linter also (pydocstyle comes to mind) in our CI framework. Of course, first step would be to first fix all the violations.

Comment (anilyil) [reply adflow/pyADflow.py]: I am going to let you two battle it out on this one. let me know if you need me to update this pr.

Comment (sseraj) [reply adflow/pyADflow.py]: The space before works locally for me but not on RTD. It would be best to follow the style guide though, so maybe we should just try to fix it on the docs build side.

Comment (nwu63) [reply adflow/pyADflow.py]: Let's talk offline and figure this out.

File: tests/reg_tests/test_actuator.py
Comment (anilyil) [reply tests/reg_tests/test_actuator.py]: As you correctly pointed out, I am dumb. I set an absolute tolerance of 0.08, and I still compare against a reference of 0. Doing it this was seemed more straightforward for me rather than explaining why the target value is a specific number. I think those type of checks should be done with a comparison to the reference.json file so that they can be trained. 

I think checking with this relatively loose tolerance is okay, because I check all functionals in the test "test_actuator_thrust_and_heat" against the reference values in https://github.com/mdolab/adflow/blob/b3469288f03f99b03032621f01990d836f137513/tests/reg_tests/refs/actuator_tests.json with `rtol=1e-12` and `atol=1e-12`. In summary, I think the tests are good to go imo if you are happy with the final modification.

Comment (anilyil) [reply tests/reg_tests/test_actuator.py]: I wanted to keep the comparison to zero just to demonstrate we expect a value of zero here, rather than some other number. Also, I test pretty much everything that touches the actuator zone so I am pretty comfortable with this individual one.

Comment (sseraj) [reply tests/reg_tests/test_actuator.py]: I am also fine with this because it basically says that the integration error should not increase past 0.08

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py:752]: Fixed by the [latest theme version ](https://github.com/mdolab/sphinx_mdolab_theme/releases/tag/v1.2.1)

File: src/solver/residuals.F90
Comment (sseraj) [reply src/solver/residuals.F90:318]: That makes sense. `LRef` is fine.

File: src/modules/actuatorRegionData.F90
Comment (sseraj) [reply src/modules/actuatorRegionData.F90]: Thanks, this looks good.

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py]: This typo is still there

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py]: The new docstring is much clearer

**Comment by sseraj** [comment]
I think the only other thing is to bump the minor version

File: adflow/pyADflow.py
Comment (anilyil) [reply adflow/pyADflow.py]: I should really learn how to read. (I thought you were referring to the not-so-clear explanation there, did not even read that word when I was editing it.) Should be fixed now.

**Comment by anilyil** [comment]
I will make the new release after this is merged right?

**Comment by sseraj** [comment]
> I will make the new release after this is merged right?

Yes, that would be great.

**Review by sseraj — APPROVED** [review]

**Review by joanibal — APPROVED** [review]

**Pull request merged**
